### PR TITLE
1. Optimize decodeBin and decodeBinsEP;  2. Optimize Intra pixel fill…

### PR DESCRIPTION
**Pull request by xrayleigh2000**
Modify point：
1.Optimize decodeBin and decodeBinsEP;
2.Optimize Intra pixel filling;
3.Optimize the SAD calculation in MDVR(5 SAD calculations at the same time to save the number of times to read data);
4.Combine BioGradFilter and pad;

Test:
1.The conformance test is performed using the command line “make test-all”, and all tests pass;
2.Tested the decoding time of single thread and multi thread. For the 0/2/10/20 thread configuration, the average decoding time is 98.06%/97.03%/97.52%/97.66% as before;

### CODE REVIEW

**Comment by K-os** [comment]
Hi xrayleigh2000,
Thank you very much for the pull request. We will look into it and give you feedback.

I only took a really short look at it, but I have one remark for you for next time: When you make multiple completely unrelated changes (e.g. the CABAC stuff and the intra prediction), can you please split it in separate commits next time. That makes reviewing the changes much easier.

**Comment by xrayleigh2000** [comment]
Hi K-os,
I divided the commit into four, the code is the same as the previous submission.

**Comment by K-os** [comment]
Oh, thank you very much.

File: source/Lib/DecoderLib/BinDecoder.cpp [hunk 1]
```diff
@@ -306,60 +313,26 @@ unsigned BinDecoder::decodeBin( unsigned ctxId )
...
-  //  Value        -= SR;
-  //  Value       <<= numBits;
-  //  Range         = LPS     << numBits;
-  //  bitsNeeded   += numBits;
-  //  if( bitsNeeded >= 0 )
-  //  {
-  //    Value      += m_Bitstream->readByte() << bitsNeeded;
-  //    bitsNeeded -= 8;
-  //  }
-  //}
```
Comment (adamjw24) [inline source/Lib/DecoderLib/BinDecoder.cpp:355]: Please don't delete comments. I like to have the unoptimized code as a documentation next to highly optimized parts, especially when doing stuff like branchless, which might be hard to read,

File: source/Lib/DecoderLib/DecLibRecon.cpp [hunk 2]
```diff
@@ -729,15 +729,18 @@ bool DecLibRecon::ctuTask( int tid, CtuTaskParam* param )
       {
         CtuData &ctuData = cs.getCtuData( ctu, line );
         ctuData.motion = decLib.m_ctuDataBufs[ctu + line * cs.pcv->widthInCtus].motion;
-        GCC_WARNING_DISABLE_class_memaccess
-        memset( ctuData.motion, 0, sizeof( CtuDataBuffers::motion ) );
-        GCC_WARNING_RESET
```
Comment (adamjw24) [inline source/Lib/DecoderLib/DecLibRecon.cpp:734]: The idea behind that is, that setting a large block of continous memory to '0' should be faster, than setting a lot of small non-continous blocks (even if only, say.. 1/4 of the actual memory needs to be reset).

Did you test this specific change for performance?

File: source/Lib/CommonLib/Contexts.h [hunk 3]
```diff
@@ -80,70 +80,55 @@ class BinProbModel : public ProbModelTables
...
   }
   static uint8_t  getRenormBitsLPS  ( unsigned LPS )                    { return    m_RenormTable_32  [LPS>>3]; }
   static uint8_t  getRenormBitsRange( unsigned range )                  { return    1; }
 private:
   uint16_t m_state[2];
-  uint8_t  m_rate;
+  uint16_t m_rate0;
+  uint16_t m_rate1;
+  uint16_t m_val0[2];
+  uint16_t m_val1[2];
```
Comment (adamjw24) [inline source/Lib/CommonLib/Contexts.h:131]: Does this have an impact? I remember testing this, without success. Doesn't hurt, tho. Is there potentially a better name than m_val0/1?

**Review by adamjw24 — CHANGES_REQUESTED** [review]
Thank you for the MR. Some stuff really looks promising.

File: source/Lib/DecoderLib/BinDecoder.cpp [hunk 4]
```diff
@@ -129,40 +133,46 @@ unsigned BinDecoder::decodeBinsEP( unsigned numBins )
   }
   unsigned remBins = numBins;
   unsigned bins    = 0;
+  int value = m_Value;
+  int range = m_Range;
+  int bitsNeeded = m_bitsNeeded;
```
Comment (adamjw24) [inline source/Lib/DecoderLib/BinDecoder.cpp:138]: Does this impact? I remember trying this, without success. I think the compiler can take care of localizing the memory in this function, no?

File: source/Lib/DecoderLib/BinDecoder.cpp (hunk 1 above)
Comment (xrayleigh2000) [reply source/Lib/DecoderLib/BinDecoder.cpp]: The unoptimized code has been added back.

File: source/Lib/DecoderLib/DecLibRecon.cpp (hunk 2 above)
Comment (xrayleigh2000) [reply source/Lib/DecoderLib/DecLibRecon.cpp:734]: In most scenes, the proportion of inter mode is larger than that of intra mode. This optimization can save a memset in inter mode.

For the 0/2/10/20 thread configuration, the average decoding time is 99.53%/99.14%/99.78%/100.06% as before;

File: source/Lib/DecoderLib/BinDecoder.cpp (hunk 4 above)
Comment (xrayleigh2000) [reply source/Lib/DecoderLib/BinDecoder.cpp:138]: After optimization, using Xcode's profile tool, the time of decodeBinsEP is reduced.

Some compilers can perform this optimization by themselves, but the optimized code is more friendly to the compiler.

File: source/Lib/CommonLib/Contexts.h (hunk 3 above)
Comment (xrayleigh2000) [reply source/Lib/CommonLib/Contexts.h]: Can m_val0/1 be modified to m_delta0/1?

Comment (adamjw24) [reply source/Lib/CommonLib/Contexts.h]: Sounds good.

**Comment by adamjw24** [comment]
Just tested the changes. It does improve the runtimes a bit. Nice, thanks. Will be merged after all discussions are resolved

**Comment by adamjw24** [comment]
Just one last thing. This is a major change with in-depth knowledge and proper impact. Feel free to add yourself to the contributors list (AUTHORS.txt) in the PR.

**Comment by xrayleigh2000** [comment]
I have added my name to AUTHORS.md

**Pull request merged**